Option in TimeUnitField using lazy_static

koenichiwa opened the pull request:
Same as #73 Just cleaned up the git mess.

Changes the internals of all TimeUnitFields to be optional. If the `TimeUnitField` specified, then its populated with an `OrdinalSet`. If it's not specified (the cron field is *) then it's None. If then the the `OrdinalSet` is requested, it will simply return a reference to a static `OrdinalSet` that is filled using `[TimeUnitField]::supported_ordinals()`. `TimeUnitField::all()` will create a TimeUnitField that will point to this static set.

TimeUnitSpec gains an extra function: `is_specified()`

I've chosen to populate this set lazily, but I had to make it statically, because the following code simply doesn't work:
```rust
fn ordinals(&self) -> &OrdinalSet {
        &self.0
        match &self.ordinals {
            Some(ordinal_set) => &ordinal_set,
            None => &self::all() // Can't return this reference, needs to be a static reference.
        }
    }
```

Lastly I cleaned up some small stuff in parsing.rs

`Seconds::from_ordinal_set(iter::once(0).collect())` becomes `Seconds::from_ordinal(0)` (etc.)

koenichiwa commented:
Decided to write some tests, found a small mistake.

zslayton commented on src/time_unit/days_of_week.rs:
```diff
@@ -2,13 +2,22 @@ use crate::error::*;
 impl TimeUnitField for DaysOfWeek {
-    fn from_ordinal_set(ordinal_set: OrdinalSet) -> Self {
-        DaysOfWeek(ordinal_set)
+    fn from_ordinal_set(ordinal_set: Option<OrdinalSet>) -> Self {
```
Accepting an `Option<_>` works fine, but to my eyes `DaysOfWeek::from_ordinal_set(None)` looks like it's using the empty set rather than all values.

Another option here: the `TimeUnitField` trait could have three methods, a low-level one that accepts an `Option<OrdinalSet>` (essentially a rename of the existing `from_ordinal_set`) and two default impls that expose a more self-documenting API for use elsewhere in the crate like `parsing.rs`:

```rust
// Each type implements this:
fn from_optional_ordinal_set(ordinal_set: Option<OrdinalSet>) -> Self;

// === The trait provides these fn impls for free ===

fn all() -> Self {
  Self::from_optional_ordinal_set(None)
}

fn from_ordinal_set(ordinal_set: OrdinalSet) -> Self {
  Self::from_optional_ordinal_set(Some(ordinal_set))
}
```

koenichiwa replied:
Yeah no problem! I'll do that.

zslayton commented on src/time_unit/mod.rs:
```diff
@@ -236,7 +237,7 @@ where
         //println!("ordinals_from_specifier for {} => {:?}", Self::name(), specifier);
         match *specifier {
-            All => Ok(Self::supported_ordinals()),
+            All => Ok(Self::supported_ordinals()),  // TODO: change to Self::ALL
```
Can you tell me about this `TODO`? Off-hand it seems like `Self::supported_ordinals()` and `Self::ALL.clone()` would do the same work. (The latter might even do more work because `lazy_static` has a small amount of overhead to dereference.)

koenichiwa replied:
Self::supported ordinals creates a new OrdinalSet. If ALL is already initialized, this should run quicker, otherwise it initializes the OrdinalSet for the next use (instead of throwing the OrdinalSet away)

zslayton replied:
`ALL` is a static ref though, and would need to be `clone()`ed to satisfy the function's return type.

koenichiwa replied:
That would defeat the purpose then. I can remove the TODO

zslayton reviewed (commented):
Why is `lazy_static` preferable to something like [`OnceCell`](https://docs.rs/once_cell/1.5.2/once_cell/index.html#lazy-initialized-global-data)? I don't have much experience with either; if there's no great answer I'm happy to move ahead with `lazy_static`.

`cron` has a lot of formatting foibles. When we reach a stopping place in the round of PRs you're working on I'd like to do a pass with `rustfmt` and add that to the CI build.

koenichiwa commented:
>Why is lazy_static preferable to something like OnceCell? I don't have much experience with either; if there's no great answer I'm happy to move ahead with lazy_static.

Never heard of OnceCell, but this is the first sentence I read on the link you send:

>This is essentially the lazy_static! macro, but without a macro.

I can try to change it to OnceCell if you prefer.

koenichiwa commented:
Im looking mor into OnceCell. It seems that there is a proposition to move it into std. Maybe it actually is better to use that instead of lazy_static.

zslayton commented on src/time_unit/mod.rs:
````diff
@@ -156,6 +156,20 @@ pub trait TimeUnitSpec {
+    /// assert_eq!(true, schedule.days_of_month().is_specified());
+    /// assert_eq!(false, schedule.months().is_specified());
+    /// ```
+    fn is_specified(&self) -> bool;
````
I have a minor quibble with the name of this function: the field _is_ specified; the user had to write `*`. How would you feel about `is_all` instead, which would return the inverse?

koenichiwa replied:
I'm a bit afraid that this implementation would break when you get to years though. If I would type Years.is_all() on 1970-2070 it should actually always return false right? Or maybe the end is actually nigh :p 

Yes you can specify 1-7 for weekdays, but then you actually did specify it, as being all weekdays.

My use case was (basically the reason why I came about this project) to change cron strings into launchd files. When you don't specify a year in a launchd file, it means that you want something to happen regardless of the year. I didn't want to print 100 or so years in my launchd file whenever I parsed a string with a * in it.

zslayton replied:
> I'm a bit afraid that this implementation would break when you get to years though. If I would type Years.is_all() on 1970-2070 it should actually always return false right? Or maybe the end is actually nigh :p

Ha, good point. The impl I laid out would be the default impl, but I think we could easily justify overriding it in the impl for `Years` to just look at `is_some()`.

> Yes you can specify 1-7 for weekdays, but then you actually did specify it, as being all weekdays.

I see the distinction you're making, but I'm not certain I want `Schedule` to make that distinction. There's not currently a distinction between `*` and `?` (for fields that accept it), and intuitively I would expect `1-7` (or whichever inclusive range) to be treated the same for fields that have a fixed range (all of them but `Years`).

> My use case was (basically the reason why I came about this project) to change cron strings into launchd files. When you don't specify a year in a launchd file, it means that you want something to happen regardless of the year. I didn't want to print 100 or so years in my launchd file whenever I parsed a string with a * in it.

Does the default impl override for `Years` address this for you?

To be clear, I could be persuaded to have an `is_specified` method, it just feels a bit hazy semantically. I wanted to see if we could get by with something that wasn't strictly bound to the input expression syntax first.

zslayton commented on src/time_unit/minutes.rs:
```diff
@@ -19,6 +28,12 @@ impl TimeUnitField for Minutes {
+        }
+    }
+    fn is_specified(&self) -> bool {
+        self.ordinals.is_some()
```
If someone wrote `0-23` for the hours field, that would be equivalent to `*` but this would return false. Is that the intended effect? What if this were something like:
```suggestion
        let max_supported_ordinals = (Self::inclusive_max() - Self::inclusive_min()) + 1;
        let is_all = self.ordinals
            .map(|o| o.len() == max_supported_ordinals)
            .unwrap_or(true);
        !is_all
```
This could be a default impl shared across all `TimeUnitField` types.

zslayton replied:
(If this were renamed from `is_specified` to `is_all` we'd obviously drop the `!` in the returned value.)

koenichiwa replied:
Thios could be an option. I feel like is_all is still a really weird function to call on years though

koenichiwa replied:
If I'd make that change, I would rather write it as:
```rust
fn is_all(&self) -> bool {
    let max_supported_ordinals = Self::inclusive_max() - Self::inclusive_min() + 1
    self.ordinals.map_or(true, |o| o.len() == max_supported_ordinals))
}
```

zslayton replied:
Yep, that's better. 👍 Didn't know about `map_or`.

koenichiwa replied:
>Didn't know about map_or.
Me neither till I tried running rustfix --pedantic. Might not be a bad idea to run that after this PR cycle.

I can do this, but ordinals() in TimeUnitField doesn't return an Option so I just changed it to 
```rust
    fn is_all(&self) -> bool {
        let max_supported_ordinals = Self::inclusive_max() - Self::inclusive_min() + 1;
        self.ordinals().len() == max_supported_ordinals as usize
    }
```
Which I'm sure is acceptable as well

koenichiwa commented on src/parsing.rs:
```diff
@@ -75,7 +75,10 @@ where
     fn from_field(field: Field) -> Result<T, Error> {
-        let mut ordinals = OrdinalSet::new(); // TODO:Combinator
+        if field.specifiers.len() == 1 && 
+            field.specifiers.get(0).unwrap() == &RootSpecifier::from(Specifier::All) 
```
Which also, if it's by itself, parses to Specifier::All

zslayton replied:
Ok, cool.

koenichiwa commented:
>Should I use lazy_static or once_cell?
To the first approximation, once_cell is both more flexible and more convenient than lazy_static and should be preferred.
Unlike once_cell, lazy_static supports spinlock-based implementation of blocking which works with #![no_std].
lazy_static has received significantly more real world testing, but once_cell is also a widely used crate

From the once_cell docs. I can change it from lazy_static without any problems. Yea or nay?

koenichiwa commented:
https://github.com/zslayton/cron/pull/76#discussion_r577742719

That would defeat the purpose then. I'll remove the TODO

zslayton commented:
> > Should I use lazy_static or once_cell?
> > To the first approximation, once_cell is both more flexible and more convenient than lazy_static and should be preferred.
> > Unlike once_cell, lazy_static supports spinlock-based implementation of blocking which works with #![no_std].
> > lazy_static has received significantly more real world testing, but once_cell is also a widely used crate
> 
> From the once_cell docs. I can change it from lazy_static without any problems. Yea or nay?

Yea. Thanks for looking into this!

koenichiwa commented:
Ok those were a couple of extra pushes, but I think I'm done now. I changed lazy_static to once_cell. I changed is_specified to is_all. Then realized it was implemented in the wrong position, so I had to change that. And lastly I changed the test string to be `"0-59 * 0-23 ?/2 1,2-4 ? *"`. I think that covers everything.

zslayton commented on src/time_unit/mod.rs:
```diff
@@ -157,7 +157,7 @@ pub trait TimeUnitSpec {
     fn count(&self) -> u32;
 
-    /// Checks if this TimeUnitSpec is specified (thus not created with a *)
+    /// Checks if this TimeUnitSpec defined as all possibilities (thus created with a '*', '?' or in the case of weekdays '1-7')
```
Nit:
```suggestion
    /// Checks if this TimeUnitSpec is defined as all possibilities (thus created with a '*', '?' or in the case of weekdays '1-7')
```

zslayton reviewed (commented):
Were you going to add an overriding impl of `is_all` for `Years` to check `is_some`? Otherwise this looks great.

koenichiwa commented:
> Were you going to add an overriding impl of `is_all` for `Years` to check `is_some`? Otherwise this looks great.

I had a bit of a problem with the name is_all in relation to years. That's why I preferred is_specified. But I do need to check in some way if * is used for years. So I'd rather leave it like this

zslayton commented:
> I had a bit of a problem with the name is_all in relation to years. That's why I preferred is_specified. But I do need to check in some way if * is used for years.

My objection to `is_specified` was/is largely a naming problem. To me, the only for a value to be _unspecified_ is for that field to be absent, which is currently only legal for `Years` (though I'd like it to be possible for `Seconds` too, #13).

Now that we have an `is_all` method that treats `?`, `*`, (e.g.) `1-7` and `<absent>` as equivalent, I think I would be ok having a separate `is_specified` method that is `false` for `<absent>` and `true` for any other expression.

Would that combination of `is_specified` and `is_all` satisfy your use case? Or do you need to be able to further distinguish between `1970-2070` and `*`, which I think is the only "all" equivalence possible in the `Years` field?

> If it is decided that the bounds of years are changed, this check will still work. So I'd rather leave it like this.

Not sure I followed this. Could you give an example?

zslayton commented:
I'm going to go ahead and merge this PR, we can resolve the outstanding discussion points in a follow-on.

Thanks again!

koenichiwa commented:
>My objection to is_specified was/is largely a naming problem. To me, the only for a value to be unspecified is for that field to be absent, which is currently only legal for Years (though I'd like it to be possible for Seconds too, #13).

Yeah my problem with is_all was largely a naming problem as well. I but I don't thionk we have to agree on every nitpick. It's your project after all.

>Not sure I followed this. Could you give an example?

The current bounds are 1970-2070, if someone comes along and decides that these bounds are insufficient. Then the is_all function on years will still work. of course it's not literaly all years, but yeah... nitpick
